[view-transitions] Skip view transition on hidden pages

nt1m opened the pull request:
#### d985c516523226282a19453a8d26b07dcfe9c119
<pre>
[view-transitions] Skip view transition on hidden pages
<a href="https://bugs.webkit.org/show_bug.cgi?id=271248">https://bugs.webkit.org/show_bug.cgi?id=271248</a>
<a href="https://rdar.apple.com/125017653">rdar://125017653</a>

Reviewed by Matt Woodrow.

Follow:
- <a href="https://drafts.csswg.org/css-view-transitions-1/#page-visibility-change-steps">https://drafts.csswg.org/css-view-transitions-1/#page-visibility-change-steps</a>
- <a href="https://github.com/w3c/csswg-drafts/issues/9543">https://github.com/w3c/csswg-drafts/issues/9543</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/transition-in-hidden-page-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/transition-in-hidden-page.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/window-resize-aborts-transition-before-ready.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/window-resize-aborts-transition.html:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations: Remove duplicate test expectations to reduce confusion.
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::visibilityStateChanged): Make this more robust to allow unregistering clients while iterating.
(WebCore::Document::resolveViewTransitionRule):
(WebCore::Document::reveal):
(WebCore::Document::startViewTransition):
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::ViewTransition):
(WebCore::ViewTransition::stop):
(WebCore::ViewTransition::visibilityStateChanged):
* Source/WebCore/dom/ViewTransition.h:

Canonical link: <a href="https://commits.webkit.org/283084@main">https://commits.webkit.org/283084@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/89117a2076e5e7a92ca060e02c9ec4c6bbd6774f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65175 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44542 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17789 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69199 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15781 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67293 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52325 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16063 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/52363 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10922 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68241 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/41180 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/56421 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32988 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37851 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13792 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14657 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/59744 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/14131 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70904 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9127 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/13609 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/59693 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9159 "Built successfully") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56481 "Exiting early after 10 failures. 84 tests run. 1 flakes") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/59968 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14371 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7547 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1220 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/40354 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41431 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/42612 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41175 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/transition-in-hidden-page-expected.txt:
```diff
@@ -0,0 +1,8 @@
+CONSOLE MESSAGE: Unhandled Promise Rejection: InvalidStateError: View transition was skipped because document visibility state is hidden.
```
Shouldn't we be concerned about this?

nt1m replied:
The test is expecting it `assert_equals(readyResult.status, "rejected");`

cdumez replied:
Seems like the test is badly written and needs fixing if we get `harness Error (FAIL)` with the expected behavior.

cdumez commented on LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/transition-in-hidden-page-expected.txt:
```diff
@@ -0,0 +1,8 @@
+CONSOLE MESSAGE: Unhandled Promise Rejection: InvalidStateError: View transition was skipped because document visibility state is hidden.
+
+Harness Error (FAIL), message = Unhandled rejection: View transition was skipped because document visibility state is hidden.
```
or this?

mattwoodrow commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -859,12 +860,26 @@ RenderViewTransitionCapture* ViewTransition::viewTransitionNewPseudoForCapturedE
+        return;
+
+    if (protectedDocument()->hidden() && protectedDocument()->activeViewTransition() == this) {
+        protectedDocument()->clearInboundViewTransitionParams();
```
Inbound view transition parameters get moved into the ViewTransition when it gets created, so they'll always be usually be nullptr already here (unless this is a same-doc VT, created before the `pagereveal` event).

I think the Document just needs to clear the params in the state change, or make the params themselves a `VisibilityChangeClient`.

mattwoodrow replied:
I can fix this in a follow-up, since it's level 2 specific.

mattwoodrow commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -859,12 +860,26 @@ RenderViewTransitionCapture* ViewTransition::viewTransitionNewPseudoForCapturedE
+        protectedDocument()->clearInboundViewTransitionParams();
+        skipViewTransition(Exception { ExceptionCode::InvalidStateError, "Skipping view transition because document visibility state has become hidden."_s });
+    } else
+        ASSERT(!protectedDocument()->activeViewTransition());
```
Looks like this fails sometimes.

My guess is that this happens because it's possible for multiple ViewTransition objects to be alive at the same time, even though only one is active. We're then running this visibility state change algorithm multiple times at once.

```suggestion
    if (protectedDocument()->hidden()) {
        if (protectedDocument()->activeViewTransition() == this)
            skipViewTransition(Exception { ExceptionCode::InvalidStateError, "Skipping view transition because document visibility state has become hidden."_s });
    } else
        ASSERT(!protectedDocument()->activeViewTransition());
```

nt1m committed:
[view-transitions] Skip view transition on hidden pages
https://bugs.webkit.org/show_bug.cgi?id=271248
rdar://125017653

Reviewed by Matt Woodrow.

Follow:
- https://drafts.csswg.org/css-view-transitions-1/#page-visibility-change-steps
- https://github.com/w3c/csswg-drafts/issues/9543

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/transition-in-hidden-page-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/transition-in-hidden-page.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/window-resize-aborts-transition-before-ready.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/window-resize-aborts-transition.html:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations: Remove duplicate test expectations to reduce confusion.
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::visibilityStateChanged): Make this more robust to allow unregistering clients while iterating.
(WebCore::Document::resolveViewTransitionRule):
(WebCore::Document::reveal):
(WebCore::Document::startViewTransition):
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::ViewTransition):
(WebCore::ViewTransition::stop):
(WebCore::ViewTransition::visibilityStateChanged):
* Source/WebCore/dom/ViewTransition.h:

Canonical link: https://commits.webkit.org/283084@main
